File: 3-week/heap-build/makeHeap.py
# python3

class HeapBuilder:
  def __init__(self):
    self._swaps = []
    self._data = []

  def ReadData(self):
    n = int(input())
    self._data = [int(s) for s in input().split()]
    self.n = n;

    assert n == len(self._data)

  def WriteResponse(self):
    print(len(self._swaps))
    for swap in self._swaps:
      print(swap[0], swap[1])

  def getLeftIndex(self, i):
    return 2 * i + 1
  

  def getRightIndex(self, i):
    return 2 * i + 2
  
  def getParent(self, i):
    return int((i - 1) / 2)

  def siftDown(self, i):
      length = len(self._data) - 1

      while self.getLeftIndex(i) <= length:
          iLeft = self.getLeftIndex(i)
          iRight = self.getRightIndex(i)
          smallest = i

          if (iLeft <= length and self._data[iLeft] < self._data[smallest]):
            smallest = iLeft

          if iRight <= length and self._data[iRight] < self._data[smallest]:
            smallest = iRight

          if i != smallest:
            current = self._data[i]

            self._data[i] = self._data[smallest]
            self._data[smallest] = current

            self._swaps.append((i, smallest))

            i = smallest
          else:
            i = iLeft

  def GenerateSwaps(self):
    for i in range(int(self.n/2), -1, -1):
      self.siftDown(i);
      

  def Solve(self):
    self.ReadData()
    self.GenerateSwaps()
    self.WriteResponse()

if __name__ == '__main__':
    heap_builder = HeapBuilder()
    heap_builder.Solve()

File: 3-week/heap-build/makeHeap.js
const rl = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false
});

function parseStr(str) {
    return str.split(/\s/);
}

function invariant(condition, message) {
    return !condition ? output(message) : true;
}

function output(message) {
    console.log(message);

    process.exit();
}

class BinaryHeapMin {
    constructor(arr) {
        let len = arr.length;
        let constraint = Math.floor(len / 2);

        this.heap = [...arr];
        this.swapsLog = [];

        for (let i = constraint; i >= 0 ; i -= 1) {
            this.siftDown(i);
        }
    }

    getLeftIndex(i) {
        return 2 * i + 1;
    }

    getRightIndex(i) {
        return 2 * i + 2;
    }

    getParent(i) {
        return Math.floor((i - 1) / 2);
    }

    insert(node) {
        let len = this.heap.length;

        this.heap[len] = node;
        this.siftUp(len);
    }

    siftUp(i) {
        while(i >= 0 && this.heap[this.getParent(i)] > this.heap[i]) {
            let iParent = this.getParent(i);
            let parent = this.heap[iParent];

            this.heap[iParent] = this.heap[i];
            this.heap[i] = parent;

            i = iParent;
        }
    }

    siftDown(i) {
        let len = this.heap.length - 1;

        while(this.getLeftIndex(i) <= len) {
            let iLeft = this.getLeftIndex(i);
            let iRight = this.getRightIndex(i);
            let smallest = i;

            if (iLeft <= len && this.heap[iLeft] < this.heap[smallest]) {
                smallest = iLeft;
            }

            if (iRight <= len && this.heap[iRight] < this.heap[smallest]) {
                smallest = iRight;
            }

            if (i !== smallest) {
                let current = this.heap[i];

                this.heap[i] = this.heap[smallest];
                this.heap[smallest] = current;

                this.swapsLog.push(`${i} ${smallest}`);

                i = smallest;
            } else {
                i = iLeft;
            }
        }
    }

}

const lines = [];

rl.on('line', (input) => {
    invariant(Boolean(input), 'input cant be blank');

    lines.push(parseStr(input.trim()));

    if (lines.length === 2) {
        let [n, arr] = lines;
        let validData = [];

        n = Number(n);

        for(let i = 0; i < n; i += 1) {
            validData[i] = Number(arr[i]);
        }

        const heap = new BinaryHeapMin(validData);

        console.log(heap.swapsLog.length);

        if (heap.swapsLog.length > 0) {
            console.log(heap.swapsLog.join('\n'));
        }

        process.exit()
    }
});
